### ISSUE TITLE
Filtering UnionArray Changes DataType

### ISSUE THREAD

**Issue by tustvold** [apache/arrow-rs#1595]
**Describe the bug**

When filtering a UnionArray the output will only include the types that had at least one value selected by the filter. I think this is a bug, as I would not expect a filter to mutate the schema of the input

**Expected behavior**

This test should pass

```
let mut builder = UnionBuilder::new_dense(4);
builder.append::<Int32Type>("A", 1).unwrap();
builder.append::<Float64Type>("B", 3.2).unwrap();
let array = builder.build().unwrap();
let filter_array = BooleanArray::from(vec![true, false]);
let c = filter(&array, &filter_array).unwrap();

assert_eq!(array.data_type(), c.data_type())
```

**Comment by tfeda** [comment]
The code runs fine for me. Perhaps it got fixed in #1589 or #1632?

**Comment by tustvold** [comment]
I can't reproduce this even on commits before those changes, so possibly a PEBCAC on my part here :sweat_smile:
